File: scripts/generate_structured_cases.js
const fs = require('fs');
const path = require('path');

function generateCases(numCases = 10, outputDir = 'data/cases') {
    if (!fs.existsSync(outputDir)) {
        fs.mkdirSync(outputDir, { recursive: true });
    }

    const firstNames = ["Arjun", "Vikram", "Rahul", "Amit", "Suresh", "Ramesh", "Karan", "Ravi", "Sanjay", "Anil"];
    const lastNames = ["Sharma", "Singh", "Verma", "Patel", "Kumar", "Gupta", "Das", "Jain", "Mehta", "Bose"];
    const locations = ["Mumbai", "Delhi", "Bangalore", "Hyderabad", "Chennai", "Kolkata", "Pune", "Ahmedabad", "Jaipur", "Surat"];
    const orgs = ["Global Tech", "Apex Holdings", "Crescent Traders", "Blue Ocean", "Pinnacle Group"];

    for (let i = 1; i <= numCases; i++) {
        const caseId = `CASE-${100 + i}`;
        const caseDir = path.join(outputDir, `case_${i.toString().padStart(2, '0')}`);
        
        if (!fs.existsSync(caseDir)) {
            fs.mkdirSync(caseDir, { recursive: true });
        }

        const suspect = `${firstNames[Math.floor(Math.random() * firstNames.length)]} ${lastNames[Math.floor(Math.random() * lastNames.length)]}`;
        let associate = `${firstNames[Math.floor(Math.random() * firstNames.length)]} ${lastNames[Math.floor(Math.random() * lastNames.length)]}`;
        while (associate === suspect) {
            associate = `${firstNames[Math.floor(Math.random() * firstNames.length)]} ${lastNames[Math.floor(Math.random() * lastNames.length)]}`;
        }

        const suspectPhone = `+91${Math.floor(Math.random() * 4000000000 + 6000000000)}`;
        const associatePhone = `+91${Math.floor(Math.random() * 4000000000 + 6000000000)}`;

        const suspectAcc = `ACC-${suspect.split(' ')[0].toUpperCase().substring(0, 3)}-${Math.floor(Math.random() * 9000 + 1000)}`;
        const associateAcc = `ACC-${associate.split(' ')[0].toUpperCase().substring(0, 3)}-${Math.floor(Math.random() * 9000 + 1000)}`;

        const location = locations[Math.floor(Math.random() * locations.length)];
        const org = orgs[Math.floor(Math.random() * orgs.length)];

        // 1. FIR.txt
        const firContent = `FIRST INFORMATION REPORT - ${caseId}
Date: 2024-03-${Math.floor(Math.random() * 16 + 10).toString().padStart(2, '0')}
Location: ${location}
Subject: Suspicious Activity

Details:
A suspicious transaction was observed involving ${suspect}.
The contact number provided is ${suspectPhone}.
The suspect is associated with an organization named ${org}.
Another associate, ${associate}, is linked to the operations.
The contact number for the associate is ${associatePhone}.
They were seen transferring funds through ${suspectAcc} and ${associateAcc}.
`;
        fs.writeFileSync(path.join(caseDir, 'FIR.txt'), firContent, 'utf8');

        // 2. CDR.csv
        const cdrData = [
            ["caller_number", "callee_number", "call_timestamp", "duration", "location"],
            [suspectPhone, associatePhone, `2024-03-14 10:${Math.floor(Math.random() * 50 + 10)}:00`, Math.floor(Math.random() * 270 + 30), location],
            [associatePhone, suspectPhone, `2024-03-15 14:${Math.floor(Math.random() * 50 + 10)}:00`, Math.floor(Math.random() * 540 + 60), location],
            [suspectPhone, `+91${Math.floor(Math.random() * 4000000000 + 6000000000)}`, `2024-03-16 09:${Math.floor(Math.random() * 50 + 10)}:00`, Math.floor(Math.random() * 110 + 10), location]
        ];
        const cdrCsv = cdrData.map(row => row.join(',')).join('\n');
        fs.writeFileSync(path.join(caseDir, 'CDR.csv'), cdrCsv, 'utf8');

        // 3. Financial.csv
        const finData = [
            ["sender_account", "receiver_account", "amount", "call_timestamp", "person_mentioned"],
            [suspectAcc, associateAcc, Math.floor(Math.random() * 450000 + 50000), "2024-03-14 11:00:00", associate],
            [associateAcc, suspectAcc, Math.floor(Math.random() * 90000 + 10000), "2024-03-15 15:00:00", suspect],
            [suspectAcc, `ACC-UNK-${Math.floor(Math.random() * 9000 + 1000)}`, Math.floor(Math.random() * 700000 + 200000), "2024-03-16 10:00:00", "Unknown Entity"]
        ];
        const finCsv = finData.map(row => row.join(',')).join('\n');
        fs.writeFileSync(path.join(caseDir, 'Financial.csv'), finCsv, 'utf8');

        // 4. metadata.json
        const metadata = {
            case_id: caseId,
            description: `Generated case data for ${suspect} and ${associate}.`
        };
        fs.writeFileSync(path.join(caseDir, 'metadata.json'), JSON.stringify(metadata, null, 4), 'utf8');
    }

    console.log(`Successfully generated ${numCases} cases in ${outputDir}`);
}

generateCases(10);

File: scripts/generate_structured_cases.py
import os
import random
import csv
import json

def generate_cases(num_cases=5, output_dir="data/cases"):
    os.makedirs(output_dir, exist_ok=True)
    
    first_names = ["Arjun", "Vikram", "Rahul", "Amit", "Suresh", "Ramesh", "Karan", "Ravi", "Sanjay", "Anil"]
    last_names = ["Sharma", "Singh", "Verma", "Patel", "Kumar", "Gupta", "Das", "Jain", "Mehta", "Bose"]
    locations = ["Mumbai", "Delhi", "Bangalore", "Hyderabad", "Chennai", "Kolkata", "Pune", "Ahmedabad", "Jaipur", "Surat"]
    orgs = ["Global Tech", "Apex Holdings", "Crescent Traders", "Blue Ocean", "Pinnacle Group"]
    
    for i in range(1, num_cases + 1):
        case_id = f"CASE-{100 + i}"
        case_dir = os.path.join(output_dir, f"case_{i:02d}")
        os.makedirs(case_dir, exist_ok=True)
        
        # Pick unique entities for this case to make it completely different
        suspect = f"{random.choice(first_names)} {random.choice(last_names)}"
        associate = f"{random.choice(first_names)} {random.choice(last_names)}"
        while associate == suspect:
            associate = f"{random.choice(first_names)} {random.choice(last_names)}"
            
        suspect_phone = f"+91{random.randint(6000000000, 9999999999)}"
        associate_phone = f"+91{random.randint(6000000000, 9999999999)}"
        
        suspect_acc = f"ACC-{suspect.split()[0].upper()[:3]}-{random.randint(1000, 9999)}"
        associate_acc = f"ACC-{associate.split()[0].upper()[:3]}-{random.randint(1000, 9999)}"
        
        location = random.choice(locations)
        org = random.choice(orgs)
        
        # 1. Generate Unstructured Text (FIR)
        fir_content = f"""FIRST INFORMATION REPORT - {case_id}
Date: 2024-03-{random.randint(10, 25):02d}
Location: {location}
Subject: Suspicious Activity

Details:
A suspicious transaction was observed involving {suspect}.
The contact number provided is {suspect_phone}.
The suspect is associated with an organization named {org}.
Another associate, {associate}, is linked to the operations.
The contact number for the associate is {associate_phone}.
They were seen transferring funds through {suspect_acc} and {associate_acc}.
"""
        with open(os.path.join(case_dir, "FIR.txt"), "w", encoding="utf-8") as f:
            f.write(fir_content)
            
        # 2. Generate Structured Data (CDR.csv)
        # Needs columns: caller_number, callee_number, call_timestamp, duration, location
        cdr_data = [
            ["caller_number", "callee_number", "call_timestamp", "duration", "location"],
            [suspect_phone, associate_phone, f"2024-03-14 10:{random.randint(10,59)}:00", random.randint(30, 300), location],
            [associate_phone, suspect_phone, f"2024-03-15 14:{random.randint(10,59)}:00", random.randint(60, 600), location],
            [suspect_phone, f"+91{random.randint(6000000000, 9999999999)}", f"2024-03-16 09:{random.randint(10,59)}:00", random.randint(10, 120), location]
        ]
        with open(os.path.join(case_dir, "CDR.csv"), "w", newline="", encoding="utf-8") as f:
            writer = csv.writer(f)
            writer.writerows(cdr_data)
            
        # 3. Generate Structured Data (Financial.csv)
        # Needs columns: sender_account, receiver_account, amount, call_timestamp, person_mentioned
        fin_data = [
            ["sender_account", "receiver_account", "amount", "call_timestamp", "person_mentioned"],
            [suspect_acc, associate_acc, random.randint(50000, 500000), "2024-03-14 11:00:00", associate],
            [associate_acc, suspect_acc, random.randint(10000, 100000), "2024-03-15 15:00:00", suspect],
            [suspect_acc, f"ACC-UNK-{random.randint(1000, 9999)}", random.randint(200000, 900000), "2024-03-16 10:00:00", "Unknown Entity"]
        ]
        with open(os.path.join(case_dir, "Financial.csv"), "w", newline="", encoding="utf-8") as f:
            writer = csv.writer(f)
            writer.writerows(fin_data)
            
        # 4. Generate metadata.json to identify it as a specific case for ingestion
        metadata = {
            "case_id": case_id,
            "description": f"Generated case data for {suspect} and {associate}."
        }
        with open(os.path.join(case_dir, "metadata.json"), "w", encoding="utf-8") as f:
            json.dump(metadata, f, indent=4)
            
    print(f"Successfully generated {num_cases} cases in {output_dir}")

if __name__ == "__main__":
    generate_cases(10)
